# Day 3 - Week 4: Hands-On Lab

## What I Learned From the Video

I watched a video about **overfitting, underfitting, and regularization** to understand these concepts better before applying them in the hands-on lab.

The main thing I learned is that getting a very high training score does not always mean that the model is good.

If the model performs very well on the training data but its performance becomes worse on new data, this means the model may be **overfitting**. It learned the training data too closely instead of learning a general pattern.

On the other hand, **underfitting** happens when the model is too simple. In this case, it cannot learn enough from the training data, so both the training and validation performance are lower.

I also learned that we can reduce overfitting by controlling the complexity of the model. For linear models, techniques such as **Ridge (L2)** and **Lasso (L1)** add a penalty to the model. Ridge reduces the size of the coefficients, while Lasso can reduce some coefficients to zero.

The strength of this penalty is controlled by `alpha`. Increasing `alpha` makes the regularization stronger and usually makes the model simpler.

### Applying What I Learned

In the hands-on lab, I applied the same idea using a **Decision Tree**.

First, I trained a Decision Tree without limiting its depth. It performed extremely well on the training data, but the validation score was lower. This helped me see overfitting in practice.

Then I created a very simple tree using:

```python
max_depth=1
```

The model could not learn enough from the data, and its performance was lower. This showed me an example of underfitting.

Finally, I reduced the complexity of the overfitted tree using:

```python
max_depth=4
min_samples_leaf=5
```

After doing this, the training score decreased a little, but the difference between the training and validation scores became smaller.

From this experiment, I understood that the goal is not to get the highest possible training accuracy. The goal is to build a model that learns the important patterns from the training data and can still perform well on new unseen data.

This helped me understand the **bias-variance trade-off** in a more practical way instead of only learning its definition.


### Importing And dataset

In [1]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

X, y = load_breast_cancer(return_X_y=True)

X_train, X_val, y_train, y_val = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

print("Training samples:", X_train.shape[0])
print("Validation samples:", X_val.shape[0])

Training samples: 455
Validation samples: 114


### Step 1: Deliberately overfit a model (e.g. a very deep decision tree) and confirm the large train-vs validation gap.

In [2]:
overfit_model = DecisionTreeClassifier(
    random_state=42
)

overfit_model.fit(X_train, y_train)

train_pred = overfit_model.predict(X_train)
val_pred = overfit_model.predict(X_val)

overfit_train_score = accuracy_score(y_train, train_pred)
overfit_val_score = accuracy_score(y_val, val_pred)

print("Training Accuracy:", overfit_train_score)
print("Validation Accuracy:", overfit_val_score)
print("Gap:", overfit_train_score - overfit_val_score)

Training Accuracy: 1.0
Validation Accuracy: 0.9122807017543859
Gap: 0.08771929824561409


## Step 1: Overfitting

I trained a Decision Tree without limiting its depth.

The model achieved a very high training accuracy, while the validation accuracy was lower.

The gap between the training and validation scores indicates that the model is overfitting.

The model learned the training data too closely and did not generalize equally well to unseen validation data.

### Step 2: Deliberately underfit a model (e.g. an overly simple one) and confirm both scores are low.

In [9]:
from sklearn.dummy import DummyClassifier

underfit_model = DummyClassifier(
    strategy="most_frequent"
)

underfit_model.fit(X_train, y_train)

train_pred = underfit_model.predict(X_train)
val_pred = underfit_model.predict(X_val)

underfit_train_score = accuracy_score(y_train, train_pred)
underfit_val_score = accuracy_score(y_val, val_pred)

print("Training Accuracy:", underfit_train_score)
print("Validation Accuracy:", underfit_val_score)
print("Gap:", underfit_train_score - underfit_val_score)

Training Accuracy: 0.6263736263736264
Validation Accuracy: 0.631578947368421
Gap: -0.005205320994794649


## Step 2: Underfitting

To create an underfitting model, I used a very simple `DummyClassifier`.

This model mainly predicts the most common class instead of learning the real patterns in the features.

The training and validation accuracy were both much lower than the Decision Tree models.

This is a clear example of underfitting because the model is too simple to learn the relationship between the features and the target.

### Step 3: Apply regularization (or reduce complexity) to the overfit model and show the gap shrink.

In [10]:
fixed_model = DecisionTreeClassifier(
    max_depth=4,
    min_samples_leaf=10,
    random_state=42
)

fixed_model.fit(X_train, y_train)

train_pred = fixed_model.predict(X_train)
val_pred = fixed_model.predict(X_val)

fixed_train_score = accuracy_score(y_train, train_pred)
fixed_val_score = accuracy_score(y_val, val_pred)

print("Training Accuracy:", fixed_train_score)
print("Validation Accuracy:", fixed_val_score)
print("Gap:", fixed_train_score - fixed_val_score)

Training Accuracy: 0.9626373626373627
Validation Accuracy: 0.9473684210526315
Gap: 0.015268941584731133


## Step 3: Fixing the Overfitting

To reduce overfitting, I limited the complexity of the Decision Tree.

I set `max_depth=4` and `min_samples_leaf=5`, which prevents the tree from growing too deep and learning very specific patterns from the training data.

After reducing the model complexity, the training accuracy decreased slightly while the validation performance remained strong.

The smaller train-validation gap indicates better generalization to unseen data.

### Step 4: Document each diagnosis and fix with the score evidence in Markdown. 

In [8]:
import pandas as pd

results = pd.DataFrame({
    "Model": [
        "Overfit Decision Tree",
        "Underfit Decision Tree",
        "Fixed Decision Tree"
    ],
    "Training Accuracy": [
        overfit_train_score,
        underfit_train_score,
        fixed_train_score
    ],
    "Validation Accuracy": [
        overfit_val_score,
        underfit_val_score,
        fixed_val_score
    ]
})

results["Gap"] = (
    results["Training Accuracy"] -
    results["Validation Accuracy"]
)

results

,Model,Training Accuracy,Validation Accuracy,Gap
0,Overfit Decision Tree,1.000000,0.912281,0.087719
1,Underfit Decision Tree,0.923077,0.921053,0.002024
2,Fixed Decision Tree,0.975824,0.921053,0.054772


### Overfit Model

- Training Accuracy: 1.0000
- Validation Accuracy: 0.9123
- Gap: 0.0877

The unrestricted Decision Tree achieved perfect training accuracy, but its validation accuracy was lower.

The relatively large gap shows that the model learned the training data too closely and was overfitting.

### Underfit Model

- Training Accuracy: 0.6263
- Validation Accuracy: 0.6315  
- Gap: -0.0052

The Dummy Classifier was intentionally very simple and did not learn the important patterns from the features.

Both training and validation accuracy were much lower than the Decision Tree models.

This is a clear example of underfitting and high bias.

### Reduced-Complexity Model

- Training Accuracy: 0.9626
- Validation Accuracy: 0.9473
- Gap: 0.0152

I reduced the complexity of the Decision Tree using `max_depth=4` and `min_samples_leaf=10`.

The training accuracy became slightly lower, but the validation accuracy improved and the train-validation gap became much smaller.

This shows that reducing model complexity helped the model generalize better to unseen data.